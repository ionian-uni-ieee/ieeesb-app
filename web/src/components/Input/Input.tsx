import React, { InputHTMLAttributes, useEffect, useRef, useState } from 'react'

interface IProps extends InputHTMLAttributes<HTMLInputElement> {
  required?: boolean
  secret?: boolean
  number?: boolean
}
const Input: React.FC<IProps> = props => {
  const {
    onChange,
    onKeyPress,
    placeholder,
    required,
    secret,
    number,
    className,
    id,
  } = props

  const [ isActive, setIsActive ] = useState<boolean>(false)
  const [ text, setText ] = useState<string>('')

  const inputRef = useRef<HTMLInputElement>(null)

  useEffect(
    () => {
      if (isActive && inputRef.current) {
        inputRef.current.focus()
      } else if (inputRef.current) {
        inputRef.current.blur()
      }
    },
    [ isActive ],
  )

  const onPlaceholderClick = () => {
    if (!isActive) {
      setIsActive(true)
    }
  }

  const onPlaceholderBlur = () => {
    const isTextEmpty = text === ''
    if (isActive && isTextEmpty) {
      setIsActive(false)
    }
  }

  const onInputChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    const eventText = e.currentTarget.value
    setText(eventText)
    const isTextEmpty = eventText === ''
    if (isTextEmpty && !isActive) {
      setIsActive(false)
    } else {
      setIsActive(true)
    }

    if (onChange) {
      onChange(e)
    }
  }

  const onInputFocus = () => {
    if (!isActive) {
      setIsActive(true)
    }
  }

  const onInputBlur = (e: React.FocusEvent<HTMLInputElement>) => {
    const eventText = e.currentTarget.value
    setText(eventText)
    const isTextEmpty = eventText === ''
    if (isActive && isTextEmpty) {
      setIsActive(false)
    }
  }

  let inputType = 'text'
  if (number) {
    inputType = 'number'
  }
  if (secret) {
    inputType = 'password'
  }

  return (
    <div
      className={`
        input
        ${isActive ? 'input_active' : ''}
        ${className}
      `}
      id={id}
    >
      <input
        className={`
          input__element
        `}
        ref={inputRef}
        required={required}
        type={inputType}
        onChange={onInputChange}
        onKeyPress={onKeyPress}
        onFocus={onInputFocus}
        onBlur={onInputBlur}
      />
      <div
        className='input__placeholder'
        onClick={onPlaceholderClick}
        onBlur={onPlaceholderBlur}
      >
        {placeholder}
      </div>
    </div>
  )
}

export default Input
