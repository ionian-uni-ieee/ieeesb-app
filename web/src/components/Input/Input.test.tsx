import React from 'react'

import { fireEvent, render } from '@testing-library/react'

import Input from './Input'

describe('Input', () => {
  it('should render', () => {
    const { container } = render(<Input />)
    expect(container).toMatchSnapshot()
  })
  it('should get an input', () => {
    const { container } = render(<Input />)
    const input = container.querySelector('input') as HTMLInputElement
    expect(input).not.toBe(null)
  })
  it('should write input', () => {
    const { container } = render(<Input />)
    const input = container.querySelector('input') as HTMLInputElement
    fireEvent.input(input, { target: { value: 'test' } })
    expect(input.value).toBe('test')
  })
})
