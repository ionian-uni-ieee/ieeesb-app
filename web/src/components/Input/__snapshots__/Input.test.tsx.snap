// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input should render 1`] = `
<div>
  <div
    class="
        input
        
        undefined
      "
  >
    <input
      class="
          input__element
        "
      type="text"
    />
    <div
      class="input__placeholder"
    />
  </div>
</div>
`;
